src/fsharp/Optimizer.fs

```diff
@@ -2543,6 +2543,37 @@ and TryDevirtualizeApplication cenv env (f,tyargs,args,m) =
                 MightMakeCriticalTailcall = false;
                 Info=UnknownValue})
 
+    // Analyze the name of the given symbol and rewrite AST to constant string expression with the name
+    | Expr.Val(vref,_,_),_,_ when valRefEq cenv.g vref cenv.g.nameof_vref ->
+        PostTypeCheckSemanticChecks.tryExtractNameOf args
+        |> Option.map (fun name ->
+            Expr.Const(Const.String name, m, cenv.g.string_ty),
+                { TotalSize = 1
+                  FunctionSize = 1
+                  HasEffect = false
+                  MightMakeCriticalTailcall = false
+                  Info = UnknownValue })
+    // Analyze the name of the given type and rewrite AST to constant string expression with the name
+    | Expr.Val(vref,_,_),_,_ when valRefEq cenv.g vref cenv.g.typenameof_vref ->
+        match tyargs with
+        | (typeName:TType):: _ ->
+            let name =
+                match typeName with 
```

Comment:
Please call ``stripTyEqns typeName``.  Consider, for example, typenameof<'T> where 'T gets later instantiated to int, e.g.

    let f(x:'T) = typenameof<'T>;  (x = 3)

This will raise a warning that ``'T`` can't be generalized and gets set to ``int``, but apart from the warning the code is allowed.

Also, please consider what happens in the presence of type aliases ``type x = int``.  I assume the intended spec is that the name of the canonical representation is returned.  Using ``stripTyEqns`` will achieve this.